backend/server.js

```diff
@@ -961,10 +955,37 @@ app.get("/qInstances/:qiid/:filename", function(req, res) {
                 if (!_(info).has("clientFiles")) {
                     return sendError(res, 500, "Question does not have clientFiles, qid: " + fileInfo.qid);
                 }
-                if (!_(info.clientFiles).contains(fileInfo.filename)) {
-                    return sendError(res, 404, "Access denied to '" + fileInfo.filename + "' for qid: " + fileInfo.qid);
+
+                if (_(info).has("clientTemplates") && _(info.clientTemplates).contains(fileInfo.filename)) {
+                    // File is template which needs to be rendered
+                    fs.readFile(path.join(fileInfo.root, fileInfo.filePath), function(err, data) {
```

Comment:
I agree that `sendFile()` won't trust relative paths, but the risk here is that `path.join()` will collapse the relative path before `sendFile()` sees it (assuming the `../` components make it past Apache, of course). I buy @kevinwang's explanation that we are checking that `fileInfo.filename` is in the `clientTemplates` whitelist, so we are all good here.